pkg/build/addimage/buildcontext.go

```diff
@@ -0,0 +1,177 @@
+/*
+Copyrigh. 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package addimage
+
+import (
+	"fmt"
+	"io"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"sigs.k8s.io/kind/pkg/build/addimage/internal/container/docker"
+	"sigs.k8s.io/kind/pkg/errors"
+	"sigs.k8s.io/kind/pkg/exec"
+	"sigs.k8s.io/kind/pkg/fs"
+	"sigs.k8s.io/kind/pkg/log"
+)
+
+// buildContext is used to build the kind node image, and contains
+// build configuration
+type buildContext struct {
+	// option fields
+	image            string
+	baseImage        string
+	additionalImages []string
+	logger           log.Logger
+	arch             string
+}
+
+// Build builds the cluster node image, the sourcedir must be set on
+// the buildContext
+func (c *buildContext) Build() (err error) {
+	return c.addImages()
+}
+
+func (c *buildContext) addImages() error {
+
+	c.logger.V(0).Info("Starting to add images to base image")
+	// pull images to local docker
+	for _, imageName := range c.additionalImages {
+		// continue if image already exists, no need to pull
+		_, err := docker.ImageID(imageName)
+		if err == nil {
+			continue
+		}
+
+		err = docker.Pull(c.logger, imageName, dockerBuildOsAndArch(c.arch), 3)
+		if err != nil {
+			c.logger.Errorf("Add image build Failed! Failed to pull image %v: %v", imageName, err)
+		}
+	}
+
+	// create build container
+	c.logger.V(0).Info("Creating build container based on " + c.baseImage)
+	// pull images to local docker
```

Comment:
we shouldn't do this anymore, it's too problematic for multi-arch, as discussed in previous review comments.

we pull straight to containerd in the current version of the node image build

Reply:
this is also a UX issue that is not thought out here, we should probably have `--pull` vs no `--pull`.

without `--pull` we should use the local docker stored image or tarball, which makes a predictable experience for locally built images

with `--pull` + `--arch` we can support multi-arch builds

`--arch` without `--pull` should possibly be prohibited or throw a warning, unless the inputs are tarballs not image names.

Reply:
supporting tarballs would make it easier to deal with architectures locally without a registry